### PR TITLE
feat(core)!: use comma separated values as a string array in the env file

### PR DESCRIPTION
**Pull request by xiaoyijun**
## Summary

- use comma separated values as a string array in the env file
- use `getEnvAsStringArray` to read `additionalConnectorPackages` from env
- add `additionalConnectorPackages` to the `env-set` and read `additionalConnectorPackages` after the `env-set` is loaded.
- update oidc private keys config and treat old envs (`OIDC_PRIVATE_KEY` and `OIDC_PRIVATE_KEY_PATHS`) as deprecated.

## Testing

Test Locally & UT passed.

### CODE REVIEW

**Comment by xiaoyijun** [comment]
We need to change the value of `OIDC_COOKIE_KEYS` to a comma-separated value before deployment.

File: packages/core/src/connectors/consts.ts [hunk 1]
```diff
@@ -16,9 +14,3 @@ const defaultPackages = [
   '@logto/connector-wechat-web',
   '@logto/connector-wechat-native',
 ];
-
-const additionalConnectorPackages = getEnv('ADDITIONAL_CONNECTOR_PACKAGES', '')
-  .split(',')
-  .filter(Boolean);
-
-export const connectorPackages = [...defaultPackages, ...additionalConnectorPackages];
```
Comment (charIeszhao) [inline packages/core/src/connectors/consts.ts:24]: Why do we remove this? Seems irrelevant

File: packages/core/src/env-set/utils.ts [hunk 2]
```diff
@@ -1,50 +1,22 @@
...
-  } catch (error: unknown) {
-    if (error instanceof InvalidStringArrayValueError) {
-      throw new EnvParseError(envKey, error.message);
-    }
-    throw error;
-  }
+  return rawValue
+    .split(',')
+    .filter(Boolean)
+    .map((value) => value.trim());
```
Comment (charIeszhao) [inline packages/core/src/env-set/utils.ts:13]: Should trim before filter

File: packages/core/src/env-set/oidc.ts [hunk 3]
```diff
@@ -26,25 +26,25 @@ const defaultLogtoOidcPrivateKey = './oidc-private-key.pem';
  * @throws An error when failed to read a private key.
  */
 export const readPrivateKeys = async (): Promise<string[]> => {
+  checkDeprecatedEnv(
```
Comment (charIeszhao) [inline packages/core/src/env-set/oidc.ts:29]: Would be better if we can check if the deprecated env exists, and only show the information when it is detected?

File: packages/core/src/env-set/oidc.ts [hunk 4]
```diff
@@ -87,18 +87,20 @@ export const readPrivateKeys = async (): Promise<string[]> => {
   const notExistPrivateKeys = privateKeyPaths.filter((path): boolean => !existsSync(path));
 
   if (notExistPrivateKeys.length > 0) {
-    const notExistPrivateKeysRawValue = JSON.stringify(notExistPrivateKeys);
     throw new Error(
-      `Private keys ${notExistPrivateKeysRawValue} configured in env \`OIDC_PRIVATE_KEY_PATHS\` not found.`
+      `Private keys ${notExistPrivateKeys.join(
+        ' and '
+      )} configured in env \`OIDC_PRIVATE_KEY_PATHS\` not found.`
```
Comment (charIeszhao) [inline packages/core/src/env-set/oidc.ts:93]: If we have more than 2 `nonExistPrivateKeys`, then the result would be `key 1 and key2 and key3`, which is not good IMO.
Better check with the `Intl.ListFormat` API? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat

File: packages/core/src/env-set/oidc.ts [hunk 5]
```diff
@@ -119,7 +121,7 @@ const readCookieKeys = async (): Promise<string[]> => {
   }
 
   const cookieKeysMissingError = new Error(
-    `The OIDC cookie keys array is missing, Please check the value of env \`${envKey}\`.`
+    `The OIDC cookie keys are not found, Please check the value of env \`${envKey}\`.`
```
Comment (charIeszhao) [inline packages/core/src/env-set/oidc.ts:124]:
```suggestion
    `The OIDC cookie keys are not found. Please check the value of env \`${envKey}\`.`
```

File: packages/core/src/env-set/utils.ts [hunk 6]
```diff
@@ -1,50 +1,22 @@
...
-
-  if (
-    !Array.isArray(values) ||
-    !values.every((value): value is string => typeof value === 'string')
-  ) {
-    throw new InvalidStringArrayValueError();
+export const checkDeprecatedEnv = (deprecatedEnv: string, newEnv: string, example: string) => {
+  if (getEnv(deprecatedEnv)) {
+    throw new Error(
+      `The env \`${deprecatedEnv}\` is deprecated, please use \`${newEnv}\` instead.\nE.g.${example}`
```
Comment (charIeszhao) [inline packages/core/src/env-set/utils.ts:19]:
```suggestion
      `The env \`${deprecatedEnv}\` is deprecated. Please use \`${newEnv}\` instead.\nE.g.${example}`
```

File: packages/core/src/env-set/oidc.ts [hunk 7]
```diff
@@ -87,18 +87,20 @@ export const readPrivateKeys = async (): Promise<string[]> => {
...
   }
 
   try {
     return privateKeyPaths.map((path): string => readFileSync(path, 'utf8'));
   } catch {
-    const privateKeyPathsRawValue = JSON.stringify(privateKeyPaths);
     throw new Error(
-      `Failed to read private keys from ${privateKeyPathsRawValue} in env \`OIDC_PRIVATE_KEY_PATHS\`.`
+      `Failed to read private keys ${privateKeyPaths.join(
+        ' and '
```
Comment (charIeszhao) [inline packages/core/src/env-set/oidc.ts:102]: Same here. Try `Intl.FormatList`

File: packages/core/src/env-set/oidc.ts [hunk 8]
```diff
@@ -26,25 +26,25 @@ const defaultLogtoOidcPrivateKey = './oidc-private-key.pem';
  * @throws An error when failed to read a private key.
  */
 export const readPrivateKeys = async (): Promise<string[]> => {
+  checkDeprecatedEnv(
+    'OIDC_PRIVATE_KEY',
+    'OIDC_PRIVATE_KEYS',
+    'OIDC_PRIVATE_KEYS=key or OIDC_PRIVATE_KEYS=key1,key2'
```
Comment (charIeszhao) [inline packages/core/src/env-set/oidc.ts:32]:
```suggestion
    'OIDC_PRIVATE_KEYS=key or OIDC_PRIVATE_KEYS=key1, key2'
```

File: packages/core/src/env-set/oidc.ts [hunk 9]
```diff
@@ -26,25 +26,25 @@ const defaultLogtoOidcPrivateKey = './oidc-private-key.pem';
...
+  );
+
+  checkDeprecatedEnv(
+    'OIDC_PRIVATE_KEY_PATH',
+    'OIDC_PRIVATE_KEY_PATHS',
+    'OIDC_PRIVATE_KEY_PATHS=path or OIDC_PRIVATE_KEY_PATHS=path1,path2'
```
Comment (charIeszhao) [inline packages/core/src/env-set/oidc.ts:38]:
```suggestion
    'OIDC_PRIVATE_KEY_PATHS=path or OIDC_PRIVATE_KEY_PATHS=path1, path2'
```

File: packages/core/src/env-set/oidc.ts (hunk 3 above)
Comment (xiaoyijun) [reply packages/core/src/env-set/oidc.ts]: Per discussion, this is a breaking change, so we don't need to warn the user.

File: packages/core/src/env-set/oidc.ts (hunk 8 above)
Comment (xiaoyijun) [reply packages/core/src/env-set/oidc.ts]: Method deleted.

File: packages/core/src/env-set/oidc.ts (hunk 9 above)
Comment (xiaoyijun) [reply packages/core/src/env-set/oidc.ts]: Method deleted.

File: packages/core/src/env-set/oidc.ts (hunk 5 above)
Comment (xiaoyijun) [reply packages/core/src/env-set/oidc.ts]: updated.

File: packages/core/src/env-set/oidc.ts (hunk 3 above)
Comment (charIeszhao) [reply packages/core/src/env-set/oidc.ts]: Let me slightly elaborate on this: It's a breaking change, and we will put the "breaking change" notice in the release notes. So there's no need to check the old config and show a friendly message any more, as we expect anyone who upgrades logto would have seen the notice beforehand.

File: packages/core/src/env-set/oidc.ts (hunk 5 above)
Comment (xiaoyijun) [reply packages/core/src/env-set/oidc.ts]: updated.

File: packages/core/src/env-set/oidc.test.ts [hunk 10]
```diff
@@ -48,63 +48,68 @@ describe('oidc env-set', () => {
     readFileSyncSpy.mockRestore();
   });
 
-  it('should read private keys if `OIDC_PRIVATE_KEY_PATH` is provided - [compatibility]', async () => {
-    // Unset the `OIDC_PRIVATE_KEY_PATHS` environment variable config by jest.setup.ts.
+  it('should generate a default OIDC private key if `OIDC_PRIVATE_KEY_PATHS` and `OIDC_PRIVATE_KEYS` are not provided', async () => {
```
Comment (charIeszhao) [inline packages/core/src/env-set/oidc.test.ts:51]:
```suggestion
  it('should generate a default OIDC private key if neither `OIDC_PRIVATE_KEY_PATHS` nor `OIDC_PRIVATE_KEYS` is provided', async () => {
```

File: packages/core/src/env-set/oidc.ts [hunk 11]
```diff
@@ -87,18 +77,20 @@ export const readPrivateKeys = async (): Promise<string[]> => {
...
   }
 
   try {
     return privateKeyPaths.map((path): string => readFileSync(path, 'utf8'));
   } catch {
-    const privateKeyPathsRawValue = JSON.stringify(privateKeyPaths);
     throw new Error(
-      `Failed to read private keys from ${privateKeyPathsRawValue} in env \`OIDC_PRIVATE_KEY_PATHS\`.`
+      `Failed to read private keys ${listFormatter.format(
+        ' and '
```
Comment (charIeszhao) [inline packages/core/src/env-set/oidc.ts:92]: `' and '` is not expected here

**Review by charIeszhao — APPROVED** [review]
Overall looks good

File: packages/core/src/env-set/oidc.test.ts [hunk 12]
```diff
@@ -48,63 +48,68 @@ describe('oidc env-set', () => {
...
+    process.env.OIDC_PRIVATE_KEYS = '';
     process.env.OIDC_PRIVATE_KEY_PATHS = '';
 
-    process.env.OIDC_PRIVATE_KEY_PATH = 'foo.pem';
+    const readFileSyncSpy = jest.spyOn(fs, 'readFileSync').mockImplementation(() => {
+      throw new Error('Intent read file error');
+    });
```
Comment (charIeszhao) [inline packages/core/src/env-set/oidc.test.ts:57]: What does this do? I don't quite get it

File: packages/core/src/connectors/index.ts [hunk 13]
```diff
@@ -19,6 +20,12 @@ const loadConnectors = async () => {
     return cachedConnectors;
   }
 
+  const {
+    values: { additionalConnectorPackages },
+  } = envSet;
+
+  const connectorPackages = [...defaultConnectorPackages, ...additionalConnectorPackages];
+
```
Comment (darcyYe) [inline packages/core/src/connectors/index.ts:28]: curious why move this part here, can't find a lot of difference.

**Review by darcyYe — APPROVED** [review]
Overall LGTM

File: packages/core/src/env-set/oidc.test.ts (hunk 12 above)
Comment (xiaoyijun) [reply packages/core/src/env-set/oidc.test.ts:57]: Mock the behavior of reading the default private key to fail.

File: packages/core/src/env-set/oidc.ts [hunk 14]
```diff
@@ -87,18 +77,20 @@ export const readPrivateKeys = async (): Promise<string[]> => {
...
+      )} from env \`OIDC_PRIVATE_KEY_PATHS\`.`
```
Comment (charIeszhao) [inline packages/core/src/env-set/oidc.ts:93]: These two error messages can be aligned

File: packages/core/src/env-set/oidc.ts [hunk 15]
```diff
@@ -87,18 +77,20 @@ export const readPrivateKeys = async (): Promise<string[]> => {
...
     );
   }
 
   try {
     return privateKeyPaths.map((path): string => readFileSync(path, 'utf8'));
   } catch {
-    const privateKeyPathsRawValue = JSON.stringify(privateKeyPaths);
     throw new Error(
-      `Failed to read private keys from ${privateKeyPathsRawValue} in env \`OIDC_PRIVATE_KEY_PATHS\`.`
+      `Failed to read private keys ${listFormatter.format(
```
Comment (charIeszhao) [inline packages/core/src/env-set/oidc.ts:91]: private key `paths`

File: packages/core/src/env-set/oidc.test.ts (hunk 12 above)
Comment (charIeszhao) [reply packages/core/src/env-set/oidc.test.ts:57]: Where is it used? And where do you assert the error message?

File: packages/core/src/env-set/oidc.ts (hunk 14 above)
Comment (xiaoyijun) [reply packages/core/src/env-set/oidc.ts:93]: They're not the same.
The first is a message for the missing file.
The second one is a failure when reading the keys.

Comment (charIeszhao) [reply packages/core/src/env-set/oidc.ts:93]: Yes, I get it. They load different things. But what I'm saying is to align the wording of the error messages. Use both `Failed to read blah blah and blah` for example

File: packages/core/src/env-set/oidc.ts (hunk 15 above)
Comment (xiaoyijun) [reply packages/core/src/env-set/oidc.ts:91]: Maybe not.
Maybe not.
```
Failed to read private keys xxx and xxx from env OIDC_PRIVATE_KEY_PATHS
```
is Ok here?

Comment (charIeszhao) [reply packages/core/src/env-set/oidc.ts:91]: Oh, I see. No problem

File: packages/core/src/connectors/index.ts (hunk 13 above)
Comment (xiaoyijun) [reply packages/core/src/connectors/index.ts:28]: We use `env-set` to manage the env values and we can only use this when the `env-set` is loaded.

File: packages/core/src/env-set/oidc.ts (hunk 14 above)
Comment (xiaoyijun) [reply packages/core/src/env-set/oidc.ts]: Removed, for we don't need to handle this exception.

File: packages/core/src/env-set/oidc.test.ts (hunk 12 above)
Comment (xiaoyijun) [reply packages/core/src/env-set/oidc.test.ts]: Per discussion, this is caused by the misleading error message. Updated.

**Review by gao-sun — APPROVED** [review]
